README.md
# Tool for Coding Tasks Specific to Civil Engineering 

Tools to make - may or may not get around to doing all (many?) of these.

General CE
- Factor of Safety Checks
- LRFD Checks
- Load Combinations (COMPLETED)
- Material Properties (steel, aluminum, concrete, plastic...)
- Simple 1d curve and multi 1d curve interpolation from charts, tables (COMPLETED)

Mechanics
- Statics
- Mechanics
- Structural Analysis

Geotech eng
- Soil classifications
- Soil Mechanics 
- Bearing Capacity
- Shear Strength
- Pressure Increase
- Lateral Pressure
- Walls Design

ASCE 7 Equations (multiple versions)
- Seismic analysis
- Wind loading
- Snow loading

AASHTO equations (multiple versions)
- Chapter 12
- Reinforced concrete
- Live loading
- Seismic loading

Other Codes
- PV1
- PV2
- NCSPA
- NDS
- ADM
- ACI
- AISC
- AISI

Structural Design
- Piles
- Footings
- Mat Foundations
- Walls
- Endwalls
- Abutments
- Relief Slabs
- Collars
- Riser Slabs

CHANGELOG.md
## CHANGELOG: ceng -- Tool for Coding Tasks Specific to Civil Engineering

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project strives to adhere to
[Semantic Versioning](http://semver.org/spec/v2.0.0.html).


### [unreleased]
   * None


### [0.5] - 2021-11-15

#### Added

  * Better interpolation documentation (`INTERP.md`)
  * Implemented API of a convenient dictionary for looking up functions that apply 
across a range of keys (`ceng.lookup.IterableDict`)

### [0.4] - 2021-10-26

#### Added

  * Implemented API for engineering interpolation (`ceng.interp`)

### [0.3.1] - 2021-10-25

#### Changed

  * corrected bug causing incorrect array shapes for load combinations

### [0.3] - 2021-10-24

#### Changed

  * improved load combination API again (breaking change)

### [0.2] - 2021-10-05

#### Changed

  * improved load combination API, added some documentation

### [0.1] - 2021-10-04

#### Features

 * Implemented API for load combinations (`ceng.load`)
 

tests/test_load.py
from inspect import signature

import numpy as np
import numpy.testing as npt
import pytest
from ceng.load import Combination, _get_identifiers, LoadCombinationExpressionError

good_combo_strs = [
    "(1.4 * D)",
    "(1.2 * D & 1.6 * L & 0.5 * (Lr | S | R))",
    "(1.2 * D & 1.6 * (S | Lr | R) & (1.0 * L | 0.5 * W))",
    "(1.2 * D & W & L & 0.5 * (Lr | S | R))",
    "(0.9 * D & W)",
    "(D)",
    "(D & L)",
    "(D & (S | Lr | R))",
    "(D & 0.75 * L & 0.75 * (Lr | S | R))",
    "(D & 0.6 * W)",
    "(D & 0.75 * L & 0.75 * 0.6 * W & 0.75 * (Lr | S | R))",
    "(0.6 * D & 0.6 * W)",
]

ok_combo_strs = [
    "1.2*D & 1.6*(Lr | S | R) & (L | 0.5*W) & 1.1*1.3*1.4*X",
    "1.2*D & 1.6*(Lr | S | R) & 0.5*(L | 0.5*W) & 1.1*1.3*1.4*X",
]

bad_combo_strs = [
    "1.2*D + 1.6*(Lr | S | R) + (L | 0.5*W)",  # plus is bad
    "1.2 * D & 1.6 * X * (S | Lr | R) & (1.0 * L | 0.5 * W)",
    "1.2 * D & X * (S | Lr | R) & (1.0 * L | 0.5 * W)",
    "1.2*D & 1.6*(Lr | S | R) & (L | 0.5*W",  # missing parens
    "1.2*D & 1.6*(Lr | S | R) & L | 0.5*W)",  # missing parens
    "1.2*D & 1.6*(Lr | S | R & (L | 0.5*W)",  # missing parens
    "1.2*D & 1.6*Lr | S | R) & (L | 0.5*W)",  # missing parens
    "1.2*D & 1.6*(Lr | S | R) & 1.1*(I(AM(A(LEAF(ON(THE(WIND) & 1.4*X & (L | 0.5*W)",  # junk
    "1.2*D & 1.6*(Lr | S | R) & (L | 0.5*W) & 1.4*X",  # dangling and operation
    "1.2*D & 1.6*(Lr | S | R) & 1.1*(1.3)*1.4*X & (L | 0.5*W) ",  # unnecessary interior parens
]


@pytest.fixture(params=range(len(good_combo_strs)), ids=good_combo_strs)
def combo_list_idx(request):
    return request.param


@pytest.fixture
def combo_str(combo_list_idx):
    return good_combo_strs[combo_list_idx]


@pytest.fixture
def combo_obj(combo_str):
    return Combination(combo_str)


def test_combo_obj(combo_obj):
    assert combo_obj


@pytest.fixture
def load_combination_func(combo_obj):
    func = eval(f"lambda {', '.join(combo_obj._identifiers)}: None")
    return combo_obj.function(func)


@pytest.fixture
def test_load_combination_func(load_combination_func):
    assert load_combination_func


combo_max_func_expected = [
    lambda D, L, Lr, S, R, W: (1.4 * D),
    lambda D, L, Lr, S, R, W: (1.2 * D + 1.6 * L + 0.5 * max(Lr, S, R)),
    lambda D, L, Lr, S, R, W: (1.2 * D + 1.6 * max(S, Lr, R) + max(1.0 * L, 0.5 * W)),
    lambda D, L, Lr, S, R, W: (1.2 * D + W + L + 0.5 * max(Lr, S, R)),
    lambda D, L, Lr, S, R, W: (0.9 * D + W),
    lambda D, L, Lr, S, R, W: (D),
    lambda D, L, Lr, S, R, W: (D + L),
    lambda D, L, Lr, S, R, W: (D + max(S, Lr, R)),
    lambda D, L, Lr, S, R, W: (D + 0.75 * L + 0.75 * max(Lr, S, R)),
    lambda D, L, Lr, S, R, W: (D + 0.6 * W),
    lambda D, L, Lr, S, R, W: (D + 0.75 * L + 0.75 * 0.6 * W + 0.75 * max(Lr, S, R)),
    lambda D, L, Lr, S, R, W: (0.6 * D + 0.6 * W),
]


@pytest.fixture(params=[
    {"D": 1, "L": 1, "Lr": 1, "S": 1, "R": 1, "W": 1},
    {"D": 1, "L": 2, "Lr": 3, "S": 4, "R": 5, "W": 6},
    {"D": -6, "L": -5, "Lr": -4, "S": -3, "R": -2, "W": -1},
], ids=["all ones", "1 to 6", "-6 to -1"])
def load_combo_input(request):
    return request.param


@pytest.fixture
def load_combo_result_expected(load_combo_input, combo_list_idx):
    return combo_max_func_expected[combo_list_idx](**load_combo_input)


def test_combo_max_result(load_combo_input, load_combination_func, load_combo_result_expected):
    s = signature(load_combination_func)
    result = load_combination_func(**{k:v for k,v in load_combo_input.items() if k in s.parameters})
    assert np.max(result) == load_combo_result_expected


@pytest.fixture
def cls_with_load_combo_method(request):
    decorator = request.param

    class C:
        @decorator
        def m(first_param, D, L):
            ...

    return C


@pytest.mark.parametrize("cls_with_load_combo_method", [
    Combination("1.4*D & 1.2*L").method,
    Combination("1.4*D & 1.2*L").classmethod,
    Combination("1.4*D & 1.2*L").staticmethod,
], ids=["method", "classmethod", "staticmethod"], indirect=True)
def test_load_combo_as_method_of_instance(cls_with_load_combo_method):
    C = cls_with_load_combo_method
    result = np.max(C().m(1, 1))
    npt.assert_almost_equal(result, np.array(2.6))


@pytest.mark.parametrize("cls_with_load_combo_method", [
    Combination("1.4*D & 1.2*L").classmethod,
    Combination("1.4*D & 1.2*L").staticmethod,
], ids=["classmethod", "staticmethod"], indirect=True)
def test_load_combo_as_method_of_cls(cls_with_load_combo_method):
    C = cls_with_load_combo_method
    result = np.max(C.m(1, 1))
    npt.assert_almost_equal(result, np.array(2.6))


def test_load_combo_as_classmethod():
    class C:
        @Combination("1.4*D & 1.2*L").classmethod
        def m(self, D, L): ...

    result = np.max(C().m(1, 1))

    npt.assert_almost_equal(result, np.array(2.6))
    npt.assert_almost_equal(result, np.array(2.6))


def test_load_combo_as_staticmethod():
    class C:
        @Combination("1.4*D & 1.2*L").method
        def m(self, D, L): ...

    result = np.max(C().m(1, 1))

    npt.assert_almost_equal(result, np.array(2.6))


def test_get_identifiers():
    test_exprs = ["GOOD_EXPRS"] + good_combo_strs + ok_combo_strs + ["BAD_EXPRS"] + bad_combo_strs

    i_identifiers = (_get_identifiers(expr) for expr in test_exprs)

    assert next(i_identifiers) == ("GOOD_EXPRS",)

    for identifiers in i_identifiers:
        assert identifiers
        if identifiers == ("BAD_EXPRS",):
            break
    while True:
        try:
            next(i_identifiers)
        except LoadCombinationExpressionError:
            continue
        except StopIteration:
            break


def test_shape_of_combination_result():
    @Combination("D & 0.75*L & 0.75*0.6*W & 0.75*(Lr | S | R)").function
    def wind_primary_load(self, D=0, L=0, W=0, Lr=0, S=0, R=0):
        ...
    D = np.array([1, 1])
    L = 2
    W = 3
    Lr = 4
    S = 5
    R = 6
    result = wind_primary_load(D, L, W, Lr, S, R)
    expected = np.array([D + 0.75*L + 0.75*0.6*W + 0.75*Lr,
                         D + 0.75*L + 0.75*0.6*W + 0.75*S,
                         D + 0.75*L + 0.75*0.6*W + 0.75*R])
    np.testing.assert_almost_equal(result, expected)


tests/test_interp.py
import pytest
import numpy as np
from ceng.interp import interp1d_twice, interp_dict


def test_documentation_example():

    rows = ("A", "B")
    sub_rows = ("1", "2")
    x = [1, 2]
    y = [10, 20, 30]
    z = {
        # TABLE HEADINGS
        #               _X=1_    |    _X=2_
        #         Y=  10 20  30  | 10  20  30
        # TABLE ROWS
        ("A", "1"): [[ 1, 2,  3], [ 4,  5,  6]],
        ("A", "2"): [[ 2, 4,  6], [ 8, 10, 12]],
        ("B", "1"): [[ 3, 6,  9], [12, 15, 18]],
        ("B", "2"): [[ 4, 8, 12], [16, 20, 24]],
    }

    xy_interp = interp_dict(x=x, y=y, z=z)
    result = xy_interp[("B", "1")](1.5, 25)
    np.testing.assert_almost_equal(result, 12.0)


@pytest.mark.parametrize("x, y, z", [
    (1.5, 1.5, 3.0),
    (1, 1.5, 1.5),
    (1.5, 1, 2.5),
])
def test_twice_interp1d_with_2d_z(x, y, z):
    f = interp1d_twice(np.array([1, 2]), np.array([1, 2, 3]), np.array([[1, 2, 3],
                                                                        [4, 5, 6]]),
                       axis=0, bounds_error=True, fill_value=None)
    assert f(x, y) == z


@pytest.mark.skip(reason="YAGNI?")
@pytest.mark.parametrize("x, y, z", [
    (1.0, 1.0, 1.0),
    (3.0, 2.0, 1.0),
    (5.0, 3.0, 1.0),
    (1.5, 1.5, 1.0),
])
def test_twice_interp1d_with_2d_x(x, y, z):
    """This is a test that might be required to pass in the future if ever a chart/table comes up demanding it.
    I hope not because it will be a real PITA.
    See:
    https://github.com/scipy/scipy/issues/14735
    """
    f = interp1d_twice(np.array([[1, 2],
                                 [3, 4],
                                 [5, 6]]), np.array([1, 2, 3]), np.array([1, 2]),
                       axis=0, bounds_error=False, fill_value=None)
    assert f(x, y) == z


tests/test_lookup.py
from ceng.lookup import IterableDict
import pytest


@pytest.fixture(name="dict_args")
def iterable_dict_arg_fixture(request):
    return request.param


@pytest.fixture(name="iterable_dict_from_kv_pairs")
def iterable_dict_from_kv_pairs_fixture(dict_args):
    return IterableDict(dict_args)


@pytest.fixture(name="iterable_dict_from_root_dict")
def iterable_dict_from_root_dict_fixture(regular_dict):
    return IterableDict(regular_dict)


@pytest.fixture(name="regular_dict")
def regular_dict_fixture(dict_args):
    return dict(dict_args)


some_keys_and_values = [
    [(range(0, 7), "spam"), (range(7, 27), "eggs")],
    [((0,), "spam"), ((7, 26), "eggs")]
]

parametrize_some_iterables = pytest.mark.parametrize("dict_args", some_keys_and_values, indirect=True)


@pytest.mark.parametrize("key, value", [
    (0, "spam"), (7, "eggs"), (26, "eggs")
])
@parametrize_some_iterables
def test_iterable_dict_getitem(iterable_dict_from_kv_pairs, iterable_dict_from_root_dict, key, value):
    """Test the getting API"""
    assert iterable_dict_from_kv_pairs[key] == value
    assert iterable_dict_from_kv_pairs.get(key) == value
    assert iterable_dict_from_root_dict[key] == value
    assert iterable_dict_from_root_dict.get(key) == value


@parametrize_some_iterables
def test_iterable_dict_api(iterable_dict_from_kv_pairs, iterable_dict_from_root_dict):
    """Test get with missing key API and test immutable"""
    assert iterable_dict_from_kv_pairs.get(-1, None) is None
    with pytest.raises(TypeError):
        iterable_dict_from_kv_pairs["spam"] = None
    assert iterable_dict_from_root_dict.get(-1, None) is None
    with pytest.raises(TypeError):
        iterable_dict_from_root_dict["spam"] = None


@parametrize_some_iterables
def test_root_dict(regular_dict, iterable_dict_from_kv_pairs, iterable_dict_from_root_dict):
    """Confirm part of API that creates/saves the root dictionary."""
    assert iterable_dict_from_kv_pairs.root == regular_dict
    assert iterable_dict_from_root_dict.root == regular_dict


def test_error_on_unhashable_iterable():
    """Be sure to mirror behavior of vanilla dict when unhashable iterables provided for keys"""
    with pytest.raises(TypeError):
        IterableDict((([1], "spam"),))


def test_error_on_incorrectly_provided_args():
    """Be sure to mirror behavior of vanilla dict when incorrect arguments provided"""
    with pytest.raises(ValueError):
        # mistakenly provide a two-tuple instead a tuple of two-tuples
        IterableDict(([1], "spam"))


ceng/common.py
import numpy as np


def flatten(maybe_seq):
    try:
        n = len(maybe_seq)
    except TypeError:
        return [maybe_seq]
    if n == 0:
        return maybe_seq
    elif n == 1:
        item = maybe_seq[0]
        try:
            return flatten(list(item))
        except TypeError:
            return [item]
    else:
        k = n // 2
        return flatten(maybe_seq[:k]) + flatten(maybe_seq[k:])


class InfoArray(np.ndarray):
    """Same as numpy array but can have an info attribute"""

    def __new__(cls, input_array, info=None):
        # Input array is an array_like
        # We first cast to be our class type
        obj = np.asarray(input_array).view(cls)
        # add the new attribute to the created instance
        obj.info = info
        # Finally, we must return the newly created object:
        return obj

    def __array_finalize__(self, obj):
        # see InfoArray.__array_finalize__ for comments
        if obj is None:
            return
        self.info = getattr(obj, 'info', None)


def iter_arg_attrs_if_attr_exists(*args, attr: str):
    """Iterate over the specified attribute of the arguments if they exist."""

    yield from (a for input in args if (a := getattr(input, attr, None)) is not None)


ceng/interp.py
from collections import defaultdict
from ceng.common import InfoArray, iter_arg_attrs_if_attr_exists
from scipy.interpolate import interp1d
import numpy as np

# Note: the scipy interp2d algorithm doesn't give the right results for twice interpolation of figures with multiple
# curves. no idea why, so have to write own interp2d algorithm.


def _interp1d_rows(inner_arr, col_arr, bounds_error, fill_value, dependent_2d):
    if dependent_2d==True:
        # z array is 2d so rows will be dependent argument to interp1d
        return [interp1d(col_arr, row, bounds_error=bounds_error, fill_value=fill_value) for row in inner_arr]
    # x or y is 2d so rows will be independent argument to interp1d
    return [interp1d(row, col_arr, bounds_error=bounds_error, fill_value=fill_value) for row in inner_arr]


def _get_row_1darr_col_1darr_and_inner_2darr(x_seq, y_seq, z_seq, first_dependent_axis):
    x_arr, y_arr, z_arr = (np.array(seq) for seq in (x_seq, y_seq, z_seq))
    x_d, y_d, z_d = (len(arr.shape) for arr in (x_arr, y_arr, z_arr))

    if not any(d==2 for d in (x_d, y_d, z_d)):
        raise ValueError("a 2d sequence or array is required")

    if all(d==1 for d in (x_d, y_d)):
        # z assumed 2d, first dependent axis is x
        row_arr, col_arr = (x_arr, y_arr) if first_dependent_axis == 0 else (y_arr, x_arr)
        inner_arr = InfoArray(z_arr, info="z")
    elif all(d==1 for d in (x_d, z_d)):
        # y assumed 2d, first dependent axis is x
        y_arr_maybe_transposed = y_arr.transpose() if first_dependent_axis == 1 else y_arr
        row_arr, col_arr = (x_arr, z_arr) if first_dependent_axis == 0 else (z_arr, x_arr)
        inner_arr = InfoArray(y_arr_maybe_transposed, info="y")
    elif all(d==1 for d in (y_d, z_d)):
        # x assumed 2d, first dependent axis is y
        x_arr_maybe_transposed = x_arr.transpose() if first_dependent_axis == 1 else x_arr
        row_arr, col_arr = (y_arr, z_arr) if first_dependent_axis == 0 else (z_arr, y_arr)
        inner_arr = InfoArray(x_arr_maybe_transposed, info="x")
    else:
        raise ValueError("two 1d sequences or arrays are required")

    if (row_arr.size, col_arr.size) != inner_arr.shape:
        raise ValueError(f"Shape of inner 2d array does not match sizes of outer arrays with first dependent at axis "
                         f"{first_dependent_axis}")

    return row_arr, col_arr, inner_arr


def _twice_interpolate_2d_using_list_of_interpolation_functions(col_v, row_v, row_arr, interp1d_rows,
                                                                bounds_error, fill_value):
    """Interpolate the function.

     Parameters
     ----------
      col_v, row_v: array_like
         column and row values to be interpolated.
     row_arr : list of interpolation functions
         values to used to create the interpolation function for second interpolation step
     interp1d_rows : list of interpolation functions
         1d interpolation functions to be used for first interpolation step

     Returns
     -------
     The interpolated value(s).
    """

    temp_x = [f(col_v) for f in interp1d_rows]
    return interp1d(row_arr, temp_x, bounds_error=bounds_error, fill_value=fill_value)(row_v)


def _twice_interp1d(x_seq, y_seq, z_seq, first_dependent_axis, bounds_error, fill_value):

    row_arr, col_arr, inner_arr = _get_row_1darr_col_1darr_and_inner_2darr(x_seq, y_seq, z_seq, first_dependent_axis)
    dependent_2d = True if inner_arr.info=="z" else False
    interp1d_rows = _interp1d_rows(inner_arr, col_arr, bounds_error, fill_value, dependent_2d=dependent_2d)

    def interpolate(x, y):

        if dependent_2d:
            axis_dict = {first_dependent_axis: x, first_dependent_axis^1: y}
            row_v, col_v = (axis_dict[i] for i in (0,1))
        else:
            row_v, col_v = y, x
        return _twice_interpolate_2d_using_list_of_interpolation_functions(col_v, row_v, row_arr, interp1d_rows,
                                                                           bounds_error, fill_value)
    return interpolate


def interp1d_twice(x, y, z, axis=0, bounds_error=True, fill_value=None):
    """Interpolate twice from many curves.

    x, y and z are values used to approximate some function f: z = f(x, y) which returns an interpolated value.

    Parameters
    ----------
    x,y : array_like of numbers
        the independent data of the curve; either both 1d, or one of x and y can be 2d if z is 1d
    z : array_like of numbers
        the dependent data of the curve; either 2d or if one of x or y are 2d, then 1d
    axis: int (0 or 1)
        axis of 2d argument assumed to correspond to the first dependent axis:
            - first dependent axis is x if either y or z is 2d
            - first dependent axis is y if x is 2d
    bounds_error : bool, optional
        if True, when interpolated values are requested outside the domain of the input data (x,y),
        a ValueError is raised. If False, then fill_value is used.
    fill_value : number, optional
        If provided, the value to use for points outside the interpolation domain. If omitted (None),
        values outside the  domain are extrapolated via nearest-neighbor extrapolation.

    Returns
    -------
    interpolator function (interpolant)
        a function that interpolates values

    Raises
    ------
    ValueError
        when unexpected or incompatible array shapes are provided
    """

    return _twice_interp1d(x, y, z, first_dependent_axis=axis, bounds_error=bounds_error, fill_value=fill_value)


def _keys_view_for_all_args_if_arg_has_keys(*args):
    """A single view of the keys of all the arguments that have them. Arguments are ignored if they don't have keys.

    Returns None if there is not at least one mapping. Raises ValueError if there are conflicting sets of keys in any
    mappings."""

    keys_iterator = (keys() for keys in iter_arg_attrs_if_attr_exists(*args, attr="keys"))
    for keys_view in keys_iterator:
        if not all(kv == keys_view for kv in keys_iterator):
            raise ValueError("mappings must have the same keys")
        break
    else:
        return None
    return keys_view


def interp_dict(x, y, z=None, axis=0, bounds_error=True, fill_value=None):
    """Produce a dictionary of interpolation functions ("interpolants").

    One or more of x,y,z must be a mapping and any mappings must have identical sets of keys.

    If z in omitted, interpolation is 1d. Otherwise it is twice 1d and exactly one of x, y, z must be 2d.

    See INTERP.md for long form example.
    """

    # get common keys view first (checks for consistency)
    keys_view = _keys_view_for_all_args_if_arg_has_keys(x, y, z)

    # turn x y z into mappings (if they aren't already)
    x_dct, y_dct, z_dct = (input if hasattr(input, "keys") else defaultdict(f)
                           for input, f in zip((x, y, z),
                                               (lambda: x,
                                                lambda: y,
                                                lambda: z)))

    # suss out the independent and dependents variables, and the interpolation method
    dependent_dct = y_dct if z is None else z_dct
    x_and_maybe_y_dicts = (x_dct,) if z is None else (x_dct, y_dct)
    interp_func = interp1d if z is None else interp1d_twice

    kwargs = dict(axis=axis, bounds_error=bounds_error, fill_value=fill_value)
    if z is None:
        if axis != 0:
            raise ValueError("axis is not used for 1d interpolation")
        del kwargs["axis"]

    result = {}
    for key in keys_view:
        x_and_maybe_y = (x_and_maybe_y_dict[key] for x_and_maybe_y_dict in x_and_maybe_y_dicts)
        result[key] = interp_func(*x_and_maybe_y, dependent_dct[key], **kwargs)

    return result


ceng/lookup.py
from collections.abc import Mapping


class IterableDict(Mapping):
    """An immutable mapping that is constructed using iterables of keys rather than individual keys.

    >>> i_dict = IterableDict((((1,2,3), "spam"), ((4,5,6), "eggs")))
    >>> i_dict
    IterableDict({(1, 2, 3): 'spam', (4, 5, 6): 'eggs'})
    >>> i_dict[1]
    'spam'
    >>> i_dict[4]
    'eggs'
    """
    root: Mapping
    _dict: dict

    def __init__(self, args):
        self.root = dict(args)
        self._dict = {k: v for keys, v in self.root.items() for k in keys}

    def __len__(self):
        return len(self._dict)

    def __iter__(self):
        yield from self._dict.keys()

    def __getitem__(self, k):
        return self._dict.__getitem__(k)

    def __repr__(self):
        return f"{self.__class__.__name__}({self.root.__repr__()})"


ceng/load.py
"""Tools for working with combinations of loads.

>>> from ceng.load Combination
>>> @Combination("1.6*D & 1.2*L & 0.5*(S | Lr | W)").function
... def my_combo(D, L, S, Lr, W): ...
...
>>> my_combo(1, 2, 3, 4, 5)
array([5.5, 6. , 6.5])

The result represents the 3 different answers from this load combination.
"""

import re
import ast
import dataclasses
from typing import TypeVar, Generic, Callable
import numpy as np
import numpy.typing as npt
import numba as nb
import wrapt

T = TypeVar('T')

_valid_operators = list("&|()")
_load_combination_re = re.compile("|".join("\\"+op for op in _valid_operators))
_multiply_re = re.compile(r"\*")


class LoadCombinationExpressionError(Exception):
    pass


def _get_identifiers(expr):
    """Parse a load combination string. Return a tuple of valid identifiers used in the string."""

    expr_lst = [s.strip() for s in _load_combination_re.split(expr) if s.strip()]
    identifiers = []

    for sub_expr in expr_lst:
        *lhs, rhs = (s.strip() for s in _multiply_re.split(sub_expr))
        # None is the case when the RHS is a parenthesized expression
        maybe_identifier = rhs if rhs else None
        for maybe_number in lhs:
            if not maybe_number.replace('.', '1').isdecimal():
                # LHS of multiplication operation has to be a number
                raise LoadCombinationExpressionError(expr, sub_expr)

        if maybe_identifier is not None:
            if not maybe_identifier.isidentifier():
                raise LoadCombinationExpressionError(expr, sub_expr)
            else:
                identifiers.append(maybe_identifier)

    try:
        ast.parse(expr)
    except Exception as e:
        raise LoadCombinationExpressionError(expr) from e

    return tuple(identifiers)


@dataclasses.dataclass
class _Factored(Generic[T]):
    """Represents a load type with a specified load factor.

    To be used in composing load expressions. E.g.:

    1.6*D & 1.2*L & 0.5*(S | Lr | W)

    Which means:

    1.6 dead load AND 1.2 live load AND 0.5 of either: snow load, live roof load, OR wind load
    """

    load_type: T
    factor: float = 1.0

    def __rmul__(self, other):
        return type(self)(self.load_type, self.factor * other)

    def __or__(self, other):
        if isinstance(other, _Factored):
            return _GroupOr((self, other))
        return NotImplemented

    def __ror__(self, other):
        return NotImplemented

    def __and__(self, other):
        if isinstance(other, _Factored):
            return _GroupAnd((self, other))
        return NotImplemented

    def __rand__(self, other):
        return NotImplemented


class _Group(tuple[_Factored]):
    """_Factored objects that have been combined"""

    @property
    def matrix(self):
        factor_arr = np.array([factored.factor for factored in self])
        if isinstance(self, _GroupOr):
            return np.diag(factor_arr)
        elif isinstance(self, _GroupAnd):
            return factor_arr.reshape((1, factor_arr.shape[0]))


class _GroupOr(_Group):
    """_Factored objects that have been `__or__`ed together

        factored_load_type_a | factored_load_type_b
    """

    def __or__(self, other):
        if isinstance(other, _Factored):
            return type(self)((*self, other))
        return NotImplemented

    def __ror__(self, other):
        return NotImplemented

    def __and__(self, other):
        return NotImplemented

    def __rand__(self, other):
        if isinstance(other, _Factored):
            other = _GroupAnd((other,))
        if isinstance(other, (_GroupOr, _GroupAnd)):
            return tuple((other, self))
        if isinstance(other, tuple):
            return tuple((*other, self))
        return NotImplemented

    def __rmul__(self, other):
        return type(self)(type(v)(v.load_type, other*v.factor) for v in self)


class _GroupAnd(_Group):
    """_Factored objects that have been `__and__`ed together

        factored_load_type_a & factored_load_type_b
    """

    def __or__(self, other):
        return  NotImplemented

    def __ror__(self, other):
        return  NotImplemented

    def __and__(self, other):
        if isinstance(other, _Factored):
            return type(self)((*self, other))
        return NotImplemented

    def __rand__(self, other):
        return  NotImplemented


@dataclasses.dataclass(frozen=True)
class Combination:
    """A callable expression of the combination of multiple loads."""

    expr: str
    matrix: npt.ArrayLike = dataclasses.field(init=False, repr=False, compare=False)

    _identifiers: list[str] = dataclasses.field(init=False, repr=False)
    _call_handler: Callable[..., npt.ArrayLike] = dataclasses.field(init=False, repr=False, compare=False)

    def __post_init__(self):
        object.__setattr__(self, "_identifiers", _get_identifiers(self.expr))
        self._init_matrix()
        self._init_call_handler()

    def __str__(self):
        return self.expr

    def __and__(self, other):
        if isinstance(other, (_GroupAnd, _GroupOr)):
            return type(self)((*self, other))
        return  NotImplemented

    def _init_matrix(self):
        """Initialize a numpy array representing the load combination."""

        ns = {k:_Factored(k) for k in self._identifiers}
        try:
            expr_eval = eval(self.expr, ns)
        except Exception as e:
            raise LoadCombinationExpressionError(self.expr) from e

        if isinstance(expr_eval, _Factored):
            group_tup = (_GroupAnd((expr_eval,)),)
        elif isinstance(expr_eval, _GroupAnd):
            group_tup = (expr_eval,)
        elif isinstance(expr_eval, tuple):
            group_tup = expr_eval
        else:
            raise LoadCombinationExpressionError(f"{self.expr!r} evaluated to type {type(expr_eval).__qualname__}")

        arr_seq = [group.matrix for group in group_tup]
        object.__setattr__(self, "matrix", _row_by_row_concatenation_of_array_seq(arr_seq))

    def _init_call_handler(self):

        # TODO: create handler using AST

        comma_sep_identifiers = ", ".join(self._identifiers)
        f_src_template = f"""
@{{decorator}}
def {{func_name}}({comma_sep_identifiers}):
    {{func_body}}"""

        ns = dict(vectorize=nb.vectorize, njit=nb.njit)
        row_funcs = list()

        for factors in self.matrix:
            # vectorize each of the row funcs to support broadcasting
            vectorized_src_format_dict = dict(
                decorator = "vectorize",
                func_name = "vectorized",
                func_body = "return " + " + ".join(f"{f}*{i}" for f, i in zip(factors, self._identifiers))
            )
            vectorized_src = f_src_template.format(**vectorized_src_format_dict)

            # jit each of the row funcs to support kwd args
            njited_src_format_dict = dict(
                decorator = "njit",
                func_name = "njited",
                func_body = f"return vectorized({comma_sep_identifiers})"
            )
            njited_src = f_src_template.format(**njited_src_format_dict)

            src = vectorized_src + "\n\n\n" + njited_src

            ns_copy = ns.copy()
            exec(src, ns_copy)
            row_funcs.append(ns_copy["njited"])

        if len(row_funcs)==1:
            row_func = row_funcs[0]

            def _call_handler(*args, **kwargs):
                return np.array(row_func(*args, **kwargs))
        else:
            def _call_handler(*args, **kwargs):
                return np.array([f(*args, **kwargs) for f in row_funcs])

        # prime the function for the float type (this seems slightly faster than providing dtype info)
        _call_handler(*(1. for _ in range(len(self._identifiers))))

        object.__setattr__(self, "_call_handler", _call_handler)

    def _decorator(self, func, inner_dec=lambda f: f):

        @wrapt.decorator
        def combine_the_loads_wrapper(wrapped, instance, args, kwargs):
            return self(*args, **kwargs)

        return combine_the_loads_wrapper(inner_dec(func))

    def function(self, func):
        """Decorator to apply to a load combination function. Automatically implements load combination.

        Example:
        >>> @Combination("1.6*D & 1.2*L & 0.5*(S | Lr | W)").function
        ... def combo(D, L, S, Lr, W):
        ...    pass
        ...

        Produces:

        >>> combo(1, 2, 3, 4, 5)
        array([5.5, 6. , 6.5])
        """

        return self._decorator(func)

    def method(self, func):
        """Decorator to apply to a load combination method. Automatically implements load combination.

        Example:
        >>> class C:
        ...     @Combination("1.6*D & 1.2*L & 0.5*(S | Lr | W)").method
        ...     def combo(self, D, L, S, Lr, W):
        ...         pass
        ...

        Produces:

        >>> C().combo(1, 2, 3, 4, 5)
        array([5.5, 6. , 6.5])
        """

        return self._decorator(func)

    def staticmethod(self, func):
        """Decorator to apply to a load combination method. Automatically implements load combination.

        Example:
        >>> class C:
        ...     @Combination("1.6*D & 1.2*L & 0.5*(S | Lr | W)").staticmethod
        ...     def combo(D, L, S, Lr, W):
        ...         pass
        ...

        Produces:

        >>> C().combo(1, 2, 3, 4, 5)
        array([5.5, 6. , 6.5])
        """

        return self._decorator(func, staticmethod)

    def classmethod(self, func):
        """Decorator to apply to a load combination method. Automatically implements load combination.

        Example:
        >>> class C:
        ...     @Combination("1.6*D & 1.2*L & 0.5*(S | Lr | W)").classmethod
        ...     def combo(cls, D, L, S, Lr, W):
        ...         pass
        ...

        Produces:

        >>> C().combo(1, 2, 3, 4, 5)
        array([5.5, 6. , 6.5])
        """

        return self._decorator(func, classmethod)

    def __call__(self, *args, **kwargs):
        return self._call_handler(*args, **kwargs)


def _row_by_row_concatenation_of_array_seq(arr_seq):
    """Combine two arrays using row by row concatenation.

    Example
    =======

    arr0:           arr1:
    [[1,2,3],       [[1,2],
    [4,5,6],        [3,4]]
    [7,8,9]]

    Into form of:

    [[1,2,3,1,2],
    [1,2,3,3,4],
    [4,5,6,1,2],
    [4,5,6,3,4],
    [7,8,9,1,2],
    [7,8,9,3,4]]
    """

    seq_rows = np.array([arr.shape[0] for arr in arr_seq])
    products = np.array([seq_rows[i:].prod() for i,_ in enumerate(seq_rows)])
    rows = seq_rows.prod()

    return np.hstack([
        np.tile(np.repeat(arr, repeats=prod//arr_rows, axis=0), (rows//prod ,1))
        for arr, arr_rows, prod in zip(arr_seq, seq_rows, products)
    ])
